connmgr/connmanager.go

```diff
@@ -0,0 +1,321 @@
+	ConnDisconnected
+	ConnFailed
+)
+
+// ConnReq is the connection request to a network address. If permanent, the
+// connection will be retried on disconnection.
+type ConnReq struct {
+	Addr      string
+	Permanent bool
+
+	id         uint64
+	conn       net.Conn
+	state      ConnState
+	stateMtx   sync.RWMutex
+	retryCount uint32
+}
+
+// ID returns a unique identifier for the connection request.
+func (c *ConnReq) ID() uint64 {
+	return atomic.LoadUint64(&c.id)
+}
+
+// State is the connection state of the requested connection.
+func (c *ConnReq) State() ConnState {
+	c.stateMtx.RLock()
+	state := c.state
+	c.stateMtx.RUnlock()
+	return state
+}
+
+// OnConnectionFunc is the signature of the callback function which is used to
+// subscribe to new connections.
+type OnConnectionFunc func(*ConnReq, net.Conn)
+
+// OnDisconnectionFunc is the signature of the callback function which is used to
+// notify disconnections.
+type OnDisconnectionFunc func(*ConnReq)
+
+var (
+	// maxRetryDuration is the max amount of time retrying of a persistent
+	// connection is allowed to grow to.  This is necessary since the retry
+	// logic uses a backoff mechanism which increases the interval base done
+	// the number of retries that have been done.
+	maxRetryDuration = time.Minute * 5
+
+	// maxFailedAttempts is the maximum number of successive failed connection
+	// attempts after which network failure is assumed and new connections will
+	// be delayed by the configured retry duration.
+	maxFailedAttempts = uint64(10)
+)
+
+// Config holds the configuration options related to the connection manager.
+type Config struct {
+	// MaxOutbound is the maximum number of outbound network connections to
+	// maintain. Defaults to 8.
+	MaxOutbound uint32
+
+	// RetryDuration is the duration to wait before retrying connection
+	// requests. Defaults to 1s.
+	RetryDuration time.Duration
+
+	// OnConnection is a callback that is fired when a new connection is
+	// established.
+	OnConnection OnConnectionFunc
+
+	// OnDisconnection is a callback that is fired when a connection is
+	// disconnected.
+	OnDisconnection OnDisconnectionFunc
+
+	// GetNewAddress is a way to get an address to make a network connection
+	// to.  If nil, no new connections will be made automatically.
+	GetNewAddress AddressFunc
+
+	// Dial connects to the address on the named network. It cannot be nil.
+	Dial DialFunc
+}
+
+// handleConnected is used to queue a successful connection.
+type handleConnected struct {
+	c    *ConnReq
+	conn net.Conn
+}
+
+// handleDisconnected is used to remove a connection.
+type handleDisconnected struct {
+	id    uint64
+	retry bool
+}
+
+// handleFailed is used to remove a pending connection.
+type handleFailed struct {
+	c   *ConnReq
+	err error
+}
+
+// ConnManager provides a manager to handle network connections.
+type ConnManager struct {
+	cfg            Config
+	wg             sync.WaitGroup
+	conns          map[uint64]*ConnReq
+	connReqCount   uint64
```

Comment:
The same deal regarding all of the fields used with `sync/atomic`.  They all need to be at the top.  I'd follow the pattern used in the `peer` package.  Namely:

```Go
	// The following variables must only be used atomically.
	connReqCount   uint64
	stop           int32
```